### PR TITLE
i-t: rewrite hooks

### PR DESCRIPTION
**Pull request by nabijaczleweli**
### Motivation and Context
A sad sight!

### Description
See commit message.

### How Has This Been Tested?
It hasn't! Will soon, time permitting, but I haven't, and if I don't open it I'll forget.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes. – none apply
- [ ] I have run the ZFS Test Suite with this change applied. – as above
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by behlendorf** [comment]
Manually merged the unrelated man page fixes from this PR.  The rewritten hooks all look sensible (and much cleaner!), but I have not tested them myself.  @nabijaczleweli your original comment says the rewrite has not been tested, please let me know when you've had a chance to test it out and are happy with the results.

**Comment by nabijaczleweli** [comment]
Rebased to include keylocation=https:// back-end. More test results below. @behlendorf this is g2g.

Sorted log diff with /etc/hostname and without libcurl is same as before ([log-new-nocurl](https://github.com/openzfs/zfs/files/6471414/log-new-nocurl.gz)), otherwise ([log-new-curl-nohostname](https://github.com/openzfs/zfs/files/6471416/log-new-curl-nohostname.gz)):
```diff
diff --git a/log-new-nocurl.s b/log-new-curl-nohostname.s
index 97967d4..0796c8c 100644
--- a/log-new-nocurl.s
+++ b/log-new-curl-nohostname.s
@@ -10,42 +10,66 @@ Adding binary /lib/x86_64-linux-gnu/libgcc_s.so.1
 Adding binary /lib/x86_64-linux-gnu/libnss_dns-2.31.so
 Adding binary /lib/x86_64-linux-gnu/libnss_files-2.31.so
 Adding binary /lib/x86_64-linux-gnu/libnss_systemd.so.2
+Adding binary /lib/x86_64-linux-gnu/librtmp.so.1
 Adding binary /lib/x86_64-linux-gnu/libselinux.so.1
+Adding binary /lib/x86_64-linux-gnu/libssl.so.1.1
 Adding binary-link /usr/lib64/ld-linux-x86-64.so.2
 Adding binary-link /usr/lib/systemd/systemd-udevd
 Adding binary-link /usr/lib/x86_64-linux-gnu/libacl.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libaudit.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libblkid.so.1
+Adding binary-link /usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1
+Adding binary-link /usr/lib/x86_64-linux-gnu/libbrotlidec.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libcap-ng.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libcap.so.2
+Adding binary-link /usr/lib/x86_64-linux-gnu/libcom_err.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/libcrypt.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libc.so.6
+Adding binary-link /usr/lib/x86_64-linux-gnu/libcurl.so.4
 Adding binary-link /usr/lib/x86_64-linux-gnu/libdl.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/libffi.so.7
 Adding binary-link /usr/lib/x86_64-linux-gnu/libgcrypt.so.20
 Adding binary-link /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0
+Adding binary-link /usr/lib/x86_64-linux-gnu/libgmp.so.10
+Adding binary-link /usr/lib/x86_64-linux-gnu/libgnutls.so.30
 Adding binary-link /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libgpg-error.so.0
+Adding binary-link /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2
+Adding binary-link /usr/lib/x86_64-linux-gnu/libhogweed.so.6
+Adding binary-link /usr/lib/x86_64-linux-gnu/libidn2.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libip4tc.so.2
+Adding binary-link /usr/lib/x86_64-linux-gnu/libk5crypto.so.3
+Adding binary-link /usr/lib/x86_64-linux-gnu/libkeyutils.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libkmod.so.2
+Adding binary-link /usr/lib/x86_64-linux-gnu/libkrb5.so.3
+Adding binary-link /usr/lib/x86_64-linux-gnu/libkrb5support.so.0
+Adding binary-link /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2
+Adding binary-link /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/liblz4.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/liblzma.so.5
 Adding binary-link /usr/lib/x86_64-linux-gnu/libmount.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libm.so.6
+Adding binary-link /usr/lib/x86_64-linux-gnu/libnettle.so.8
+Adding binary-link /usr/lib/x86_64-linux-gnu/libnghttp2.so.14
 Adding binary-link /usr/lib/x86_64-linux-gnu/libnss_dns.so
 Adding binary-link /usr/lib/x86_64-linux-gnu/libnss_dns.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/libnss_files.so
 Adding binary-link /usr/lib/x86_64-linux-gnu/libnss_files.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/libnvpair.so.3
+Adding binary-link /usr/lib/x86_64-linux-gnu/libp11-kit.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libpam.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libpcre2-8.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libpcre.so.3
+Adding binary-link /usr/lib/x86_64-linux-gnu/libpsl.so.5
 Adding binary-link /usr/lib/x86_64-linux-gnu/libpthread.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libresolv.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/librt.so.1
+Adding binary-link /usr/lib/x86_64-linux-gnu/libsasl2.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/libseccomp.so.2
+Adding binary-link /usr/lib/x86_64-linux-gnu/libssh2.so.1
+Adding binary-link /usr/lib/x86_64-linux-gnu/libtasn1.so.6
 Adding binary-link /usr/lib/x86_64-linux-gnu/libtspi.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libtss2-esys.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libtss2-mu.so.0
@@ -59,6 +83,7 @@ Adding binary-link /usr/lib/x86_64-linux-gnu/libtss2-tcti-swtpm.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libtss2-tcti-tabrmd.so
 Adding binary-link /usr/lib/x86_64-linux-gnu/libtss2-tcti-tabrmd.so.0
 Adding binary-link /usr/lib/x86_64-linux-gnu/libudev.so.1
+Adding binary-link /usr/lib/x86_64-linux-gnu/libunistring.so.2
 Adding binary-link /usr/lib/x86_64-linux-gnu/libuuid.so.1
 Adding binary-link /usr/lib/x86_64-linux-gnu/libuutil.so.3
 Adding binary-link /usr/lib/x86_64-linux-gnu/libzfs_core.so.3
@@ -88,32 +113,54 @@ Adding binary /usr/lib/x86_64-linux-gnu/ld-2.31.so
 Adding binary /usr/lib/x86_64-linux-gnu/libacl.so.1.1.2253
 Adding binary /usr/lib/x86_64-linux-gnu/libaudit.so.1.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libblkid.so.1.1.0
+Adding binary /usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1.0.9
+Adding binary /usr/lib/x86_64-linux-gnu/libbrotlidec.so.1.0.9
 Adding binary /usr/lib/x86_64-linux-gnu/libc-2.31.so
 Adding binary /usr/lib/x86_64-linux-gnu/libcap-ng.so.0.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libcap.so.2.44
+Adding binary /usr/lib/x86_64-linux-gnu/libcom_err.so.2.1
 Adding binary /usr/lib/x86_64-linux-gnu/libcrypt.so.1.1.0
+Adding binary /usr/lib/x86_64-linux-gnu/libcurl.so.4.7.0
 Adding binary /usr/lib/x86_64-linux-gnu/libdl-2.31.so
 Adding binary /usr/lib/x86_64-linux-gnu/libffi.so.7.1.0
 Adding binary /usr/lib/x86_64-linux-gnu/libgcrypt.so.20.2.8
 Adding binary /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0.6600.8
 Adding binary /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.6600.8
 Adding binary /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0.6600.8
+Adding binary /usr/lib/x86_64-linux-gnu/libgmp.so.10.4.1
+Adding binary /usr/lib/x86_64-linux-gnu/libgnutls.so.30.29.1
 Adding binary /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6600.8
 Adding binary /usr/lib/x86_64-linux-gnu/libgpg-error.so.0.29.0
+Adding binary /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2
+Adding binary /usr/lib/x86_64-linux-gnu/libhogweed.so.6.1
+Adding binary /usr/lib/x86_64-linux-gnu/libidn2.so.0.3.7
 Adding binary /usr/lib/x86_64-linux-gnu/libip4tc.so.2.0.0
+Adding binary /usr/lib/x86_64-linux-gnu/libk5crypto.so.3.1
+Adding binary /usr/lib/x86_64-linux-gnu/libkeyutils.so.1.9
 Adding binary /usr/lib/x86_64-linux-gnu/libkmod.so.2.3.6
+Adding binary /usr/lib/x86_64-linux-gnu/libkrb5.so.3.3
+Adding binary /usr/lib/x86_64-linux-gnu/libkrb5support.so.0.1
+Adding binary /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2.11.5
+Adding binary /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2.11.5
 Adding binary /usr/lib/x86_64-linux-gnu/liblz4.so.1.9.3
 Adding binary /usr/lib/x86_64-linux-gnu/liblzma.so.5.2.5
 Adding binary /usr/lib/x86_64-linux-gnu/libm-2.31.so
 Adding binary /usr/lib/x86_64-linux-gnu/libmount.so.1.1.0
+Adding binary /usr/lib/x86_64-linux-gnu/libnettle.so.8.1
+Adding binary /usr/lib/x86_64-linux-gnu/libnghttp2.so.14.20.1
 Adding binary /usr/lib/x86_64-linux-gnu/libnvpair.so.3.0.0
+Adding binary /usr/lib/x86_64-linux-gnu/libp11-kit.so.0.3.0
 Adding binary /usr/lib/x86_64-linux-gnu/libpam.so.0.85.1
 Adding binary /usr/lib/x86_64-linux-gnu/libpcre2-8.so.0.10.1
 Adding binary /usr/lib/x86_64-linux-gnu/libpcre.so.3.13.3
+Adding binary /usr/lib/x86_64-linux-gnu/libpsl.so.5.3.2
 Adding binary /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
 Adding binary /usr/lib/x86_64-linux-gnu/libresolv-2.31.so
 Adding binary /usr/lib/x86_64-linux-gnu/librt-2.31.so
+Adding binary /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25
 Adding binary /usr/lib/x86_64-linux-gnu/libseccomp.so.2.5.1
+Adding binary /usr/lib/x86_64-linux-gnu/libssh2.so.1.0.1
+Adding binary /usr/lib/x86_64-linux-gnu/libtasn1.so.6.6.0
 Adding binary /usr/lib/x86_64-linux-gnu/libtspi.so.1.2.0
 Adding binary /usr/lib/x86_64-linux-gnu/libtss2-esys.so.0.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libtss2-mu.so.0.0.0
@@ -126,6 +173,7 @@ Adding binary /usr/lib/x86_64-linux-gnu/libtss2-tcti-mssim.so.0.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libtss2-tcti-swtpm.so.0.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libtss2-tcti-tabrmd.so.0.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0
+Adding binary /usr/lib/x86_64-linux-gnu/libunistring.so.2.1.0
 Adding binary /usr/lib/x86_64-linux-gnu/libuuid.so.1.3.0
 Adding binary /usr/lib/x86_64-linux-gnu/libuutil.so.3.0.0
 Adding binary /usr/lib/x86_64-linux-gnu/libzfs_core.so.3.0.0
@@ -138,7 +186,6 @@ Adding config /etc/default/zfs
 Adding config /etc/gai.conf
 Adding config /etc/host.conf
 Adding config /etc/hostid
-Adding config /etc/hostname
 Adding config /etc/hosts
 Adding config /etc/modprobe.d/dkms.conf
 Adding config /etc/nsswitch.conf
@@ -148,6 +195,7 @@ Adding config /etc/zfs/zfs-functions
 Adding config /lib/modprobe.d/aliases.conf
 Adding config /lib/modprobe.d/fbdev-blacklist.conf
 Adding config /lib/modprobe.d/systemd.conf
+Adding config /tmp/hostname.68EH2LUoFV
 Adding config /usr/share/initramfs-tools/conf.d/zfs
 Adding firmware /lib/firmware/cis/DP83903.cis
 Adding firmware /lib/firmware/cis/LA-PCM.cis
```

The listing diff is as expected ([id-nc.gz](https://github.com/openzfs/zfs/files/6471422/id-nc), [id-cnh](https://github.com/openzfs/zfs/files/6471424/id-cnh.gz)):
```diff
diff --git a/id-nc b/id-cnh
index 71ffe0b..24f0256 100644
--- a/id-nc
+++ b/id-cnh
@@ -1164,15 +1164,23 @@ usr/lib/x86_64-linux-gnu/libaudit.so.1
 usr/lib/x86_64-linux-gnu/libaudit.so.1.0.0
 usr/lib/x86_64-linux-gnu/libblkid.so.1
 usr/lib/x86_64-linux-gnu/libblkid.so.1.1.0
+usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1
+usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1.0.9
+usr/lib/x86_64-linux-gnu/libbrotlidec.so.1
+usr/lib/x86_64-linux-gnu/libbrotlidec.so.1.0.9
 usr/lib/x86_64-linux-gnu/libc-2.31.so
 usr/lib/x86_64-linux-gnu/libc.so.6
 usr/lib/x86_64-linux-gnu/libcap-ng.so.0
 usr/lib/x86_64-linux-gnu/libcap-ng.so.0.0.0
 usr/lib/x86_64-linux-gnu/libcap.so.2
 usr/lib/x86_64-linux-gnu/libcap.so.2.44
+usr/lib/x86_64-linux-gnu/libcom_err.so.2
+usr/lib/x86_64-linux-gnu/libcom_err.so.2.1
 usr/lib/x86_64-linux-gnu/libcrypt.so.1
 usr/lib/x86_64-linux-gnu/libcrypt.so.1.1.0
 usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
+usr/lib/x86_64-linux-gnu/libcurl.so.4
+usr/lib/x86_64-linux-gnu/libcurl.so.4.7.0
 usr/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
 usr/lib/x86_64-linux-gnu/libdl-2.31.so
 usr/lib/x86_64-linux-gnu/libdl.so.2
@@ -1187,14 +1195,36 @@ usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
 usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.6600.8
 usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0
 usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0.6600.8
+usr/lib/x86_64-linux-gnu/libgmp.so.10
+usr/lib/x86_64-linux-gnu/libgmp.so.10.4.1
+usr/lib/x86_64-linux-gnu/libgnutls.so.30
+usr/lib/x86_64-linux-gnu/libgnutls.so.30.29.1
 usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
 usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6600.8
 usr/lib/x86_64-linux-gnu/libgpg-error.so.0
 usr/lib/x86_64-linux-gnu/libgpg-error.so.0.29.0
+usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2
+usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2
+usr/lib/x86_64-linux-gnu/libhogweed.so.6
+usr/lib/x86_64-linux-gnu/libhogweed.so.6.1
+usr/lib/x86_64-linux-gnu/libidn2.so.0
+usr/lib/x86_64-linux-gnu/libidn2.so.0.3.7
 usr/lib/x86_64-linux-gnu/libip4tc.so.2
 usr/lib/x86_64-linux-gnu/libip4tc.so.2.0.0
+usr/lib/x86_64-linux-gnu/libk5crypto.so.3
+usr/lib/x86_64-linux-gnu/libk5crypto.so.3.1
+usr/lib/x86_64-linux-gnu/libkeyutils.so.1
+usr/lib/x86_64-linux-gnu/libkeyutils.so.1.9
 usr/lib/x86_64-linux-gnu/libkmod.so.2
 usr/lib/x86_64-linux-gnu/libkmod.so.2.3.6
+usr/lib/x86_64-linux-gnu/libkrb5.so.3
+usr/lib/x86_64-linux-gnu/libkrb5.so.3.3
+usr/lib/x86_64-linux-gnu/libkrb5support.so.0
+usr/lib/x86_64-linux-gnu/libkrb5support.so.0.1
+usr/lib/x86_64-linux-gnu/liblber-2.4.so.2
+usr/lib/x86_64-linux-gnu/liblber-2.4.so.2.11.5
+usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2
+usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2.11.5
 usr/lib/x86_64-linux-gnu/liblz4.so.1
 usr/lib/x86_64-linux-gnu/liblz4.so.1.9.3
 usr/lib/x86_64-linux-gnu/liblzma.so.5
@@ -1203,6 +1233,10 @@ usr/lib/x86_64-linux-gnu/libm-2.31.so
 usr/lib/x86_64-linux-gnu/libm.so.6
 usr/lib/x86_64-linux-gnu/libmount.so.1
 usr/lib/x86_64-linux-gnu/libmount.so.1.1.0
+usr/lib/x86_64-linux-gnu/libnettle.so.8
+usr/lib/x86_64-linux-gnu/libnettle.so.8.1
+usr/lib/x86_64-linux-gnu/libnghttp2.so.14
+usr/lib/x86_64-linux-gnu/libnghttp2.so.14.20.1
 usr/lib/x86_64-linux-gnu/libnss_dns-2.31.so
 usr/lib/x86_64-linux-gnu/libnss_dns.so
 usr/lib/x86_64-linux-gnu/libnss_dns.so.2
@@ -1212,21 +1246,33 @@ usr/lib/x86_64-linux-gnu/libnss_files.so.2
 usr/lib/x86_64-linux-gnu/libnss_systemd.so.2
 usr/lib/x86_64-linux-gnu/libnvpair.so.3
 usr/lib/x86_64-linux-gnu/libnvpair.so.3.0.0
+usr/lib/x86_64-linux-gnu/libp11-kit.so.0
+usr/lib/x86_64-linux-gnu/libp11-kit.so.0.3.0
 usr/lib/x86_64-linux-gnu/libpam.so.0
 usr/lib/x86_64-linux-gnu/libpam.so.0.85.1
 usr/lib/x86_64-linux-gnu/libpcre.so.3
 usr/lib/x86_64-linux-gnu/libpcre.so.3.13.3
 usr/lib/x86_64-linux-gnu/libpcre2-8.so.0
 usr/lib/x86_64-linux-gnu/libpcre2-8.so.0.10.1
+usr/lib/x86_64-linux-gnu/libpsl.so.5
+usr/lib/x86_64-linux-gnu/libpsl.so.5.3.2
 usr/lib/x86_64-linux-gnu/libpthread-2.31.so
 usr/lib/x86_64-linux-gnu/libpthread.so.0
 usr/lib/x86_64-linux-gnu/libresolv-2.31.so
 usr/lib/x86_64-linux-gnu/libresolv.so.2
 usr/lib/x86_64-linux-gnu/librt-2.31.so
 usr/lib/x86_64-linux-gnu/librt.so.1
+usr/lib/x86_64-linux-gnu/librtmp.so.1
+usr/lib/x86_64-linux-gnu/libsasl2.so.2
+usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25
 usr/lib/x86_64-linux-gnu/libseccomp.so.2
 usr/lib/x86_64-linux-gnu/libseccomp.so.2.5.1
 usr/lib/x86_64-linux-gnu/libselinux.so.1
+usr/lib/x86_64-linux-gnu/libssh2.so.1
+usr/lib/x86_64-linux-gnu/libssh2.so.1.0.1
+usr/lib/x86_64-linux-gnu/libssl.so.1.1
+usr/lib/x86_64-linux-gnu/libtasn1.so.6
+usr/lib/x86_64-linux-gnu/libtasn1.so.6.6.0
 usr/lib/x86_64-linux-gnu/libtspi.so.1
 usr/lib/x86_64-linux-gnu/libtspi.so.1.2.0
 usr/lib/x86_64-linux-gnu/libtss2-esys.so.0
@@ -1252,6 +1298,8 @@ usr/lib/x86_64-linux-gnu/libtss2-tctildr.so.0
 usr/lib/x86_64-linux-gnu/libtss2-tctildr.so.0.0.0
 usr/lib/x86_64-linux-gnu/libudev.so.1
 usr/lib/x86_64-linux-gnu/libudev.so.1.7.0
+usr/lib/x86_64-linux-gnu/libunistring.so.2
+usr/lib/x86_64-linux-gnu/libunistring.so.2.1.0
 usr/lib/x86_64-linux-gnu/libuuid.so.1
 usr/lib/x86_64-linux-gnu/libuuid.so.1.3.0
 usr/lib/x86_64-linux-gnu/libuutil.so.3
```

Likewise the content diff
```
$ git diff --stat new-nocurl/ new-curl-nohostname/
 {new-nocurl => new-curl-nohostname}/etc/ld.so.cache   | Bin 4634 -> 6574 bytes
 .../usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1     |   1 +
 .../usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1.0.9 | Bin 0 -> 137400 bytes
 .../usr/lib/x86_64-linux-gnu/libbrotlidec.so.1        |   1 +
 .../usr/lib/x86_64-linux-gnu/libbrotlidec.so.1.0.9    | Bin 0 -> 51384 bytes
 .../usr/lib/x86_64-linux-gnu/libcom_err.so.2          |   1 +
 .../usr/lib/x86_64-linux-gnu/libcom_err.so.2.1        | Bin 0 -> 18344 bytes
 .../usr/lib/x86_64-linux-gnu/libcurl.so.4             |   1 +
 .../usr/lib/x86_64-linux-gnu/libcurl.so.4.7.0         | Bin 0 -> 628232 bytes
 .../usr/lib/x86_64-linux-gnu/libgmp.so.10             |   1 +
 .../usr/lib/x86_64-linux-gnu/libgmp.so.10.4.1         | Bin 0 -> 525152 bytes
 .../usr/lib/x86_64-linux-gnu/libgnutls.so.30          |   1 +
 .../usr/lib/x86_64-linux-gnu/libgnutls.so.30.29.1     | Bin 0 -> 2082456 bytes
 .../usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2      |   1 +
 .../usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2    | Bin 0 -> 334616 bytes
 .../usr/lib/x86_64-linux-gnu/libhogweed.so.6          |   1 +
 .../usr/lib/x86_64-linux-gnu/libhogweed.so.6.1        | Bin 0 -> 294816 bytes
 .../usr/lib/x86_64-linux-gnu/libidn2.so.0             |   1 +
 .../usr/lib/x86_64-linux-gnu/libidn2.so.0.3.7         | Bin 0 -> 128944 bytes
 .../usr/lib/x86_64-linux-gnu/libk5crypto.so.3         |   1 +
 .../usr/lib/x86_64-linux-gnu/libk5crypto.so.3.1       | Bin 0 -> 187288 bytes
 .../usr/lib/x86_64-linux-gnu/libkeyutils.so.1         |   1 +
 .../usr/lib/x86_64-linux-gnu/libkeyutils.so.1.9       | Bin 0 -> 22448 bytes
 .../usr/lib/x86_64-linux-gnu/libkrb5.so.3             |   1 +
 .../usr/lib/x86_64-linux-gnu/libkrb5.so.3.3           | Bin 0 -> 888448 bytes
 .../usr/lib/x86_64-linux-gnu/libkrb5support.so.0      |   1 +
 .../usr/lib/x86_64-linux-gnu/libkrb5support.so.0.1    | Bin 0 -> 56352 bytes
 .../usr/lib/x86_64-linux-gnu/liblber-2.4.so.2         |   1 +
 .../usr/lib/x86_64-linux-gnu/liblber-2.4.so.2.11.5    | Bin 0 -> 63672 bytes
 .../usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2       |   1 +
 .../usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2.11.5  | Bin 0 -> 339280 bytes
 .../usr/lib/x86_64-linux-gnu/libnettle.so.8           |   1 +
 .../usr/lib/x86_64-linux-gnu/libnettle.so.8.1         | Bin 0 -> 290416 bytes
 .../usr/lib/x86_64-linux-gnu/libnghttp2.so.14         |   1 +
 .../usr/lib/x86_64-linux-gnu/libnghttp2.so.14.20.1    | Bin 0 -> 178640 bytes
 .../usr/lib/x86_64-linux-gnu/libp11-kit.so.0          |   1 +
 .../usr/lib/x86_64-linux-gnu/libp11-kit.so.0.3.0      | Bin 0 -> 1257280 bytes
 .../usr/lib/x86_64-linux-gnu/libpsl.so.5              |   1 +
 .../usr/lib/x86_64-linux-gnu/libpsl.so.5.3.2          | Bin 0 -> 75616 bytes
 .../usr/lib/x86_64-linux-gnu/librtmp.so.1             | Bin 0 -> 122280 bytes
 .../usr/lib/x86_64-linux-gnu/libsasl2.so.2            |   1 +
 .../usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25       | Bin 0 -> 113392 bytes
 .../usr/lib/x86_64-linux-gnu/libssh2.so.1             |   1 +
 .../usr/lib/x86_64-linux-gnu/libssh2.so.1.0.1         | Bin 0 -> 211272 bytes
 .../usr/lib/x86_64-linux-gnu/libssl.so.1.1            | Bin 0 -> 597792 bytes
 .../usr/lib/x86_64-linux-gnu/libtasn1.so.6            |   1 +
 .../usr/lib/x86_64-linux-gnu/libtasn1.so.6.6.0        | Bin 0 -> 83968 bytes
 .../usr/lib/x86_64-linux-gnu/libunistring.so.2        |   1 +
 .../usr/lib/x86_64-linux-gnu/libunistring.so.2.1.0    | Bin 0 -> 1574952 bytes
 49 files changed, 23 insertions(+)
```

The hostname in the no-/etc/hostname case is correct.

File: contrib/initramfs/hooks/zfs.in [hunk 1]
```diff
@@ -3,115 +3,54 @@
...
+copy_file rule   "@udevruledir@/60-zvol.rules"
+copy_file rule   "@udevruledir@/69-vdev.rules"
 
-for ii in $COPY_FILE_LIST
-do
-	dir=$(dirname "$ii")
-	[ -d "$dir" ] && mkdir -p "$DESTDIR/$dir"
-	[ -f "$ii" ] && cp -p "$ii" "$DESTDIR/$ii"
-done
+manual_add_modules zlib_deflate spl zavl zcommon znvpair zunicode zlua zfs icp
```
Comment (behlendorf) [inline contrib/initramfs/hooks/zfs.in:47]: It looks like the `zzstd` module needs to be added here.  I imagine it must have been getting automatically pulled in as a dependency somehow for this to work.

Comment (nabijaczleweli) [reply contrib/initramfs/hooks/zfs.in:47]: I just blindly copied it, but if you wanna touch this, I'd actually say we want to just say "zfs" here: `manual_add_modules()` just asks modprobe for dependencies (i.e. `modprobe --all --set-version="$version" --quiet --show-depends $MODS`, roughly, to paraphrase `hook-functions`, and that's the case as far back as *jessie*): this is to say that it is, *by definition*, correct. If it weren't, the modules wouldn't've loaded into the live system.

I don't know why that comment said that "automatic dependency resolution is unreliable on many systems" but it cannot physically be right – I assume it's folk wisdom from twenty years ago when modular kernels were new and scary.

`dracut-install -m` (as called by `instmods`) does an equivalent thing (but via libkmod and very weirdly, normal day in dracut land) – you will notice that dracut's script installs all the modules explicitly as well, and also doesn't list zzstd (though I'd hesitate to call it "obviously correct" like is the case here).

Comment (behlendorf) [reply contrib/initramfs/hooks/zfs.in]: Yup, let's try and do the right thing here by just listing `zfs` and letting depmod do its thing.  Thanks for updating it.

**Review by behlendorf — APPROVED** [review]

**Pull request merged**